### PR TITLE
feat: add `__copy__`, `__deepcopy__`, and `__replace__` for FFI objects

### PR DESCRIPTION
**Pull request by junrushao**
## Summary
This PR adds first-class copy support to TVM-FFI objects, including Python `copy.copy`, `copy.deepcopy`, and dataclass-style `__replace__`.

## Changes
- Adds reflection-based deep-copy runtime support in C++:
  - new `ffi.DeepCopy` entrypoint (`src/ffi/extra/deep_copy.cc`), which
    does memoized graph copy that preserves shared references and
    cycles.
    - recursive copy of `ffi.Array` / `ffi.Map`
    - treats `Str`/`Bytes` as immutable terminal values
    - resolves fields by runtime value so `Any`/`ObjectRef` fields containing containers/objects are copied correctly
- Auto-registers shallow-copy support for copy-constructible reflected object types:
    - introduces `__ffi_shallow_copy__` registration in `ObjectDef`
    - exposes shallow-copy method via type attributes for generic lookup
- Wires Python class registration to install:
    - `__copy__`, `__deepcopy__`, and `__replace__` for supported types
    - clear `TypeError` for non-copyable types
    - deepcopy support for container roots (`ffi.Array`, `ffi.Map`) via `ffi.DeepCopy`
- Updates dataclass/c_class method attachment to use reflected callables consistently and hook copy setup.
- Adds extensive test coverage in `tests/python/test_copy.py`, including:
    - shallow/deep copy behavior
    - shared-reference and cycle preservation
    - long-string deep copy edge cases
    - `Any` and `ObjectRef` deep copy edge cases
    - non-copyable type failures (`TestNonCopyable`)

### CODE REVIEW

File: python/tvm_ffi/registry.py [hunk 1]
```diff
@@ -335,25 +335,97 @@ def _add_class_attrs(type_cls: type, type_info: TypeInfo) -> type:
...
+
+
+def _get_deep_copy_func() -> core.Function:
+    global _deep_copy_func_cache  # noqa: PLW0603
+    if _deep_copy_func_cache is None:
+        _deep_copy_func_cache = get_global_func("ffi.DeepCopy")
+    return _deep_copy_func_cache
+
+
+_deep_copy_func_cache: core.Function | None = None
```
Comment (junrushao) [inline python/tvm_ffi/registry.py]: probably don't need this - just call `get_global_func` every time

File: include/tvm/ffi/extra/copy.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*!
+ * \file tvm/ffi/extra/copy.h
+ * \brief Reflection-based object copy utilities
+ */
+#ifndef TVM_FFI_EXTRA_COPY_H_
+#define TVM_FFI_EXTRA_COPY_H_
```
Comment (tqchen) [inline include/tvm/ffi/extra/copy.h]: deep_copy.h?

File: include/tvm/ffi/extra/copy.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ * Recursively copies the value and all reachable objects in its object graph.
+ * Objects must have copy support enabled via `refl::enable_copy()` to be deep copied.
+ * Primitive types, strings, and bytes are returned as-is (they are immutable).
+ * Arrays and Maps are recursively deep copied.
+ * Objects without copy support are kept as shared references.
+ *
+ * \param value The value to deep copy.
+ * \return The deep copied value.
+ */
+TVM_FFI_EXTRA_CXX_API Any DeepCopyValue(const Any& value);
```
Comment (tqchen) [inline include/tvm/ffi/extra/copy.h]: I feel DeepCopy as an API name is fine here

File: include/tvm/ffi/reflection/registry.h [hunk 4]
```diff
@@ -460,6 +460,49 @@ struct init {
...
+ *
+ * \code{.cpp}
+ * refl::ObjectDef<MyObject>()
+ *     .def(refl::enable_copy())
+ *     .def(refl::init<int64_t, std::string>());
+ * \endcode
+ *
+ * \note The object type must have a copy constructor.
+ */
+struct enable_copy {
```
Comment (tqchen) [inline include/tvm/ffi/reflection/registry.h]: should it be enable_deep_copy to be explicit?

File: src/ffi/extra/copy.cc [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+
+namespace tvm {
+namespace ffi {
+
+/*!
+ * \brief Deep copier that recursively copies objects using shallow copy + field replacement.
+ *
+ * Algorithm:
+ * Phase 1: Walk the entire object graph (objects, arrays, maps) using BFS,
+ *          collecting all reachable nodes in a vector. An unordered_set
```
Comment (tqchen) [inline src/ffi/extra/copy.cc]: Likely this algorithm can be simplified to a single pass algorithm with memo_ that maps previously mapped Object into a new one.

The RunCopy will internally look up the memo and return copied objects if found. If not, it will then attempt to run the deep copy recursively and store the result into the memo. There is no need to record things like node order, visited will be part of the memo_ map that dedup double visits

File: include/tvm/ffi/reflection/registry.h (hunk 4 above)
Comment (junrushao) [reply include/tvm/ffi/reflection/registry.h]: No it's for shallow copy

Comment (tqchen) [reply include/tvm/ffi/reflection/registry.h]: yes, i understand, seems the confusion part is that it enables a shallow copy API that was required for deep copy?

**Comment by junrushao** [comment]
Still - implication of cyclic reference and CoW containers (Array, Map) makes it pretty challenging to have an absolutely correct implementation of `__deepcopy__`. Converting it to draft.

File: include/tvm/ffi/reflection/registry.h (hunk 4 above)
Comment (junrushao) [reply include/tvm/ffi/reflection/registry.h]: alternatively, let's enable copy by default for all object types which are copy_constructible

File: include/tvm/ffi/extra/copy.h (hunk 3 above)
Comment (junrushao) [reply include/tvm/ffi/extra/copy.h]: good idea!

File: include/tvm/ffi/extra/copy.h (hunk 2 above)
Comment (junrushao) [reply include/tvm/ffi/extra/copy.h]: updated

**Comment by junrushao** [comment]
Updated per discussion with TQ. @tqchen please re-review

File: src/ffi/extra/deep_copy.cc [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    if (value.type_index() < TypeIndex::kTVMFFIStaticObjectBegin) {
+      return value;
+    }
+    const Object* obj = value.as<Object>();
+    if (auto it = copy_map_.find(obj); it != copy_map_.end()) {
+      return it->second;
+    }
+    int32_t ti = obj->type_index();
+    // Strings and bytes are immutable — return as-is.
+    if (ti == TypeIndex::kTVMFFIStr || ti == TypeIndex::kTVMFFIBytes) {
```
Comment (tqchen) [inline src/ffi/extra/deep_copy.cc]: this also applies to kTVMFFIShape which is immutable and can return as it is

File: include/tvm/ffi/reflection/registry.h [hunk 7]
```diff
@@ -460,6 +460,9 @@ struct init {
   }
 };
 
+/*! \brief Method name used for shallow copy registration. */
+inline constexpr const char* kShallowCopyMethodName = "__ffi_shallow_copy__";
```
Comment (tqchen) [inline include/tvm/ffi/reflection/registry.h]: kInitMethodName is member of ObjectDef, for consistency maybe make it also member? 

Alterntaively, consider group these constants into struct or namespace like `struct StaticTypeKey {`. Maybe TypeAttrName or `reflection::type_attr::`
